pkg/graveler/combined_iterator.go

```diff
@@ -67,35 +77,26 @@ func (c *CombinedIterator) advanceInnerIterators() bool {
 }
 
 func (c *CombinedIterator) Next() bool {
-	for {
-		if !c.advanceInnerIterators() {
-			return false
-		}
-		// set c.p to be the next (smaller) value or continue in case of tombstone
-		valA := c.iterA.Value()
-		valB := c.iterB.Value()
-		switch {
-		case valA == nil && valB == nil:
-			c.p = c.iterA
-			return false
-		case valA == nil:
-			c.p = c.iterB
-		case valB == nil:
-			c.p = c.iterA
-		case bytes.Equal(valA.Key, valB.Key) && valA.IsTombstone():
-			// continue without tombstone error
-			continue
-		case bytes.Compare(valA.Key, valB.Key) <= 0:
-			c.p = c.iterA
-		default:
-			c.p = c.iterB
-		}
-		if c.p.Value().IsTombstone() {
-			logging.Default().Error("unexpected tombstone")
-			continue
-		}
-		return true
+	if !c.advanceInnerIterators() {
+		return false
+	}
+	// set c.p to be the next (smaller) value
+	valA := c.iterA.Value()
+	valB := c.iterB.Value()
+	switch {
+	case valA == nil && valB == nil:
+		c.p = c.iterA
+		return false
+	case valA == nil:
+		c.p = c.iterB
+	case valB == nil:
+		c.p = c.iterA
```

Comment:
We need to skip the nil and still process the iterators based on key ordering. So in case one of the iterators points to a nil value we need to skip until we get to a value or marked as ended by setting it to nil.

Reply:
^ can ignore I thought it was the value that hold nil for tomstone